src/browser/navigation-panel.js

```diff
@@ -46,78 +46,61 @@ define((require, exports, module) => {
       })
     ]));
 
-  const updateSuggestionsSelection = (delta, suggestionsCursor, webViewerCursor) => {
-    const size = suggestionsCursor.get('list').size;
-    let value = suggestionsCursor.get('selectedIndex') + delta;
-    if (value >= size) {
-      value = -1;
-    }
-    value = Math.max(-1, value);
-    suggestionsCursor.set('selectedIndex', value);
+  // Clears selection in the suggestions
+  const unselect = suggestions =>
+    suggestions.set('selectedIndex', -1);
+
+  // Selects suggestion `n` items away relative to currently seleceted suggestion.
+  // Selection over suggestion entries is moved in a loop although there is extra
+  // "no selection" entry between last and first suggestions. Given `n` can be negative
+  // or positive in order to select suggestion before or after the current one.
+  const selectRelative = curry((n, suggestions) =>
+    suggestions.update('selectedIndex', from => {
+      const first = 0;
+      const last = suggestions.get('list').count() - 1;
+      const to = from + n;
+
+      return to > last ? -1 :
+             to < first ? -1 :
+             to;
+    }));
+
+  // Selects next / previous suggestion.
+  const selectPrevious = selectRelative(-1);
+  const selectNext = selectRelative(1);
+
+  // Returns currently selected suggestion or void if there's none.
+  const selected = suggestions => {
+    const index = suggestions.get('selectedIndex');
+    return index >= 0 ? suggestions.getIn(['list', index, 'text']) : void(0);
```

Comment:
Please not that while `x.get(a).get(b)` may throw `x.getIn([a, b])` won't it will just short circuit on with `undefined` if somewhere along the path is, there for I removed try catch.